chalk-solve/src/infer/unify.rs

```diff
@@ -57,53 +63,68 @@ impl<'t, I: Interner> Unifier<'t, I> {
         let n_b = self.table.normalize_ty_shallow(interner, b);
         let a = n_a.as_ref().unwrap_or(a);
         let b = n_b.as_ref().unwrap_or(b);
 
-        debug_span!("unify_ty_ty", ?a, ?b);
+        debug_span!("relate_ty_ty", ?variance, ?a, ?b);
 
         match (a.data(interner), b.data(interner)) {
-            // Unifying two inference variables: unify them in the underlying
-            // ena table.
+            // Relating two inference variables:
+            // If `Invariant`, unify them in the underlying ena table.
+            // If `Covariant` or `Contravariant`, push `SubtypeGoal`
             (
                 &TyData::InferenceVar(var1, kind1),
                 &TyData::InferenceVar(var2, kind2),
             ) => {
-                if kind1 == kind2 {
-                    self.unify_var_var(var1, var2)
-                } else if kind1 == TyKind::General {
-                    self.unify_general_var_specific_ty(var1, b.clone())
-                } else if kind2 == TyKind::General {
-                    self.unify_general_var_specific_ty(var2, a.clone())
-                } else {
-                    debug!(
-                        "Tried to unify mis-matching inference variables: {:?} and {:?}",
-                        kind1, kind2
-                    );
-                    Err(NoSolution)
+                match variance {
+                    Variance::Invariant => {
+                        if kind1 == kind2 {
+                            self.unify_var_var(var1, var2)
+                        } else if kind1 == TyKind::General {
+                            self.unify_general_var_specific_ty(var1, b.clone())
+                        } else if kind2 == TyKind::General {
+                            self.unify_general_var_specific_ty(var2, a.clone())
+                        } else {
+                            debug!(
+                                "Tried to unify mis-matching inference variables: {:?} and {:?}",
+                                kind1, kind2
+                            );
+                            Err(NoSolution)
+                        }
+                    },
+                    Variance::Covariant => {
+                        self.push_subtype_goal(a.clone().cast(interner), b.clone().cast(interner));
+                        Ok(())
+                    },
+                    Variance::Contravariant => {
+                        self.push_subtype_goal(b.clone().cast(interner), a.clone().cast(interner));
+                        Ok(())
+                    },
                 }
             }
 
-            // Unifying an inference variable with a non-inference variable.
+            // Relating an inference variable with a non-inference variable.
             (&TyData::InferenceVar(var, kind), ty_data @ &TyData::Apply(_))
             | (&TyData::InferenceVar(var, kind), ty_data @ &TyData::Placeholder(_))
             | (&TyData::InferenceVar(var, kind), ty_data @ &TyData::Dyn(_))
```

Comment:
I haven't updated this. I'm not sure if there's a way to do this *without* generalization?

Reply:
I added a FIXME